src/core/model.jl

```diff
@@ -107,27 +107,34 @@ export random_agent, nagents, allagents, allids, nextid, seed!
 """
     model[id]
     getindex(model::ABM, id::Integer)
-
 Return an agent given its ID.
 """
 Base.getindex(m::ABM, id::Integer) = m.agents[id]
 
 """
     model[id] = agent
     setindex!(model::ABM, agent::AbstractAgent, id::Int)
-
 Add an `agent` to the `model` at a given index: `id`.
 Note this method will return an error if the `id` requested is not equal to `agent.id`.
 **Internal method, use [`add_agents!`](@ref) instead to actually add an agent.**
 """
-function Base.setindex!(m::ABM, a::AbstractAgent, id::Int)
+function Base.setindex!(m::ABM{S,A,Dict{Int,A}}, a::AbstractAgent, id::Int) where {S,A}
     a.id ≠ id &&
-    throw(ArgumentError("You are adding an agent to an ID not equal with the agent's ID!"))
+        throw(ArgumentError("You are adding an agent to an ID not equal with the agent's ID!"))
     m.agents[id] = a
     m.maxid[] < id && (m.maxid[] += 1)
     return a
 end
 
+function Base.setindex!(m::ABM{S,A,Vector{A}}, a::AbstractAgent, id::Int) where {S,A}
+    a.id ≠ id &&
+        throw(ArgumentError("You are adding an agent to an ID not equal with the agent's ID!"))
+    id == nagents(m) + 1 || error("Cannot add agent of ID $(id) to a vector ABM of $(nagents(m)) agents. Expected ID == $(nagents(m)+1).")
+    m.maxid[] < id && (m.maxid[] += 1)
```

Comment:
Wouldn't this prevent calling `setindex!` on existing agent IDs? E.g. If there are 4 agents, with ids from 1 to 4, calling `abm[3] = MyAgent(...)` would error

Reply:
Yeah I clearly didn't think this one through properly.

Reply:
`abm[3] = MyAgent(...)` must always error. It is not supported syntax!!!! This is not the API way to add agents!!! Only the `add_agent!` function is supported!!!!

Reply:
How does one edit agents in an ABM if `abm[id] = MyAgent(id, ...)` is disallowed?

Reply:
Only `add_agent!` and `kill_agent!` can be used to add or remove agents from the model. If you want to "edit" an agent simply edit its fields in place. `a = model[id]; a.field = value`. You can't just assign a new agent to an ID previouslyt occupied by another agent.